### DISCUSSION TITLE
Deep package filtering (ignore dependency scripts)

### DISCUSSION THREAD

**Discussion by mattidupre** [vercel/turborepo#7453] [Ideas]
### Which project is this feature idea for?

Turborepo

### Describe the feature you'd like to request

When executing `turbo run` I am wanting to fully ignore packages in my workspace, including in dependencies. Unless I am mistaken, the `--filter` CLI flag will skip matches in the initial scope but won't skip those matches when they are dependencies.

**My use case:**

I am using [Yalc](https://github.com/wclr/yalc) to handle a local linked dependency. It's great about copying that linked dependency into my monorepo's `./yalc` folder on every push and getting around React's peer dependency issues. I have `./yalc/*` glob in that monorepo's workspace glob and everything else works perfect. However, when I run a Turborepo pipeline, it will try and execute that linked package's build scripts. Since it doesn't have `devDependencies`, they will inevitably fail.

### Describe the solution you'd like

**Given:**
* Pipeline `dev` depends on `build`. 
* Packages `bar` and `baz` are dependencies `foo`.
* Running `turbo run dev --filter-deep=baz`

**Desired outcome:**
* `bar/package.json:build` will execute.
* `baz/package.json:build` will NOT execute.

### Describe alternatives you've considered

I am currently using [clean-publish](https://github.com/shashkovdanil/clean-publish) in my linked package's build pipeline to strip the package.json scripts before yalc pushes. 

`"rm -rf ./publish && clean-publish --without-publish --temp-dir publish && yalc push --changed publish && rm -rf ./publish"`

This is causing some secondary headaches I am trying to avoid.

**Comment by mehulkar** [comment]
A good starting point here would be to contribute an integration test with Prysk (and docs) that confirms the behavior you’ve mentioned. I’m pretty sure you’re right, but the behavior is not documented anywhere. Are you able to do thaf @mattidupre ?

**Comment by mehulkar** [comment]
> I am using [Yalc](https://github.com/wclr/yalc) to handle a local linked dependency

Is this dependency outside of your turborepo repo? If so, the best solution may just be for us to recommend that you move this dependency into your monorepo.

> I have `./yalc/*` glob in that monorepo's workspace glob and everything else works perfect.

Curious, why are you including this as a workspace if it's generated output?

> `--filter` CLI flag will skip matches in the initial scope but won't skip those matches when they are dependencies.

This is a potential solution, but I'd like to understand the use case a bit more. (btw, for what it's worth `--filter` is to _include_ the thing, not _exclude_ the thing. `--filter=foo` only runs foo, rather than excludes `foo`. So your example is backwards)

**Comment by mattidupre** [comment]
Appreciate the response! Sorry, it's been a busy couple weeks and haven't had a chance to look into the Turbo tests and documentation yet.

Good call on my example, I neglected to include some negation. `--filter=!foo`

The linked dependency is required across multiple repos and cannot currently be published. Being able to make changes to that dependency that update in another dependent's onSave speeds up development. npm, Yarn, and pnpm have the `link` command but that usually fails to properly merge dependencies (i.e., React, which fails with multiple imported instances). Yalc exists to manually copy that dependency on every file save so the dependent's package manager can treat it as a workspace and not worry about symlinks.

Other solutions:
* a PR with Yalc adding an option to strip all scripts from the dependency's package.json on every save ("push").
* renaming dependent's scripts so they aren't captured by dependent's turbo pipelines.

It's an edge case but I'm surprised this hasn't come up for repos wanting to gradually migrate to Turbo. I did just see a similar issue is being tracked with nx [here.](https://github.com/nrwl/nx/issues/18053)
